src/Common/tests/System/Net/Http/LoopbackProxyServer.cs

```diff
@@ -218,41 +240,53 @@ private async Task ProcessConnectMethod(NetworkStream clientStream, string remot
             {
                 try
                 {
-                    byte[] buffer = new byte[8000];
-                    int bytesRead;
-                    while ((bytesRead = await clientStream.ReadAsync(buffer, 0, buffer.Length).ConfigureAwait(false)) > 0)
-                    {
-                        await serverStream.WriteAsync(buffer, 0, bytesRead).ConfigureAwait(false);
-                    }
-                    serverStream.Flush();
+                    await clientStream.CopyToAsync(serverStream).ConfigureAwait(false);
                     serverSocket.Shutdown(SocketShutdown.Send);
                 }
-                catch (IOException)
+                catch (IOException ex) when (IsRudeDisconnect(ex))
                 {
                     // Ignore rude disconnects from either side.
+                    // Close sockets to abort the other task.
+                    clientStream.Close();
+                    serverStream.Close();
+                }
+                catch
+                {
+                    clientStream.Close();
+                    serverStream.Close();
+                    throw;
                 }
             });
+
             Task serverCopyTask = Task.Run(async () =>
             {
                 try
                 {
-                    byte[] buffer = new byte[8000];
-                    int bytesRead;
-                    while ((bytesRead = await serverStream.ReadAsync(buffer, 0, buffer.Length).ConfigureAwait(false)) > 0)
-                    {
-                        await clientStream.WriteAsync(buffer, 0, bytesRead).ConfigureAwait(false);
-                    }
-                    clientStream.Flush();
+                    await serverStream.CopyToAsync(clientStream).ConfigureAwait(false);
+                    clientSocket.Shutdown(SocketShutdown.Send);
                 }
-                catch (IOException)
+                catch (IOException ex) when (IsRudeDisconnect(ex))
                 {
                     // Ignore rude disconnects from either side.
+                    // Close sockets to abort the other task.
+                    clientStream.Close();
+                    serverStream.Close();
+                }
+                catch
+                {
+                    clientStream.Close();
+                    serverStream.Close();
+                    throw;
                 }
             });
 
-            await Task.WhenAny(new[] { clientCopyTask, serverCopyTask }).ConfigureAwait(false);
+            await Task.WhenAll(new[] { clientCopyTask, serverCopyTask }).ConfigureAwait(false);
```

Comment:
Is there any situation where the client or server copy task failing could cause the other one to hang indefinitely?  I've not looked to see whether that's a possibility here, but it's often the case when two tasks are related in a client/server fashion, and it's why we have the WhenAllOrAnyFailed helper (https://github.com/dotnet/corefx/blob/0a9a366e290aded3e9bf5d082d6beee5ff560177/src/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs#L64).

Reply:
Closing the streams disposes of the sockets causing an OperationAborted error to be thrown in the other task, so I don't think this is a problem here.